cirq-core/cirq/devices/device.py

```diff
@@ -253,3 +263,108 @@ def _from_json_dict_(cls, qubits, nx_graph, **kwargs):
+                everywhere on the device.
+            gate_durations: Optional dictionary of `cirq.Gateset`
+                instances mapping to `cirq.Duration` instances for
+                gate timing metadata information. If provided,
+                must match all entries in supported_gates.
+
+        Raises:
+            ValueError: if the union of gateset keys in gate_durations,
+                do not represent an identical gateset to supported_gates.
+        """
+        qubit_pairs = list(qubit_pairs)
+        flat_pairs = [q for pair in qubit_pairs for q in pair]
+        # Keep lexigraphically smaller tuples for undirected edges.
+        sorted_pairs = sorted(qubit_pairs)
+        qubit_set = set()
+        for a, b in sorted_pairs:
+            if (b, a) not in qubit_set:
+                qubit_set.add((a, b))
+
+        connectivity = nx.Graph()
+        connectivity.add_edges_from(sorted(list(qubit_set)), directed=False)
+        super().__init__(flat_pairs, connectivity)
+        self._qubit_pairs = frozenset(qubit_set)
+        self._supported_gates = supported_gates
+
+        if gate_durations is not None:
+            working_gatefamilies = frozenset(
+                g for gset in gate_durations.keys() for g in gset.gates
+            )
+            if working_gatefamilies != supported_gates.gates:
+                overlap = working_gatefamilies.difference(supported_gates.gates)
+                raise ValueError(
+                    "Supplied gate_durations contains gates not present"
+                    f" in supported_gates. {overlap} in supported_gates"
+                    " is False."
+                )
+
+        self._gate_durations = gate_durations
+
+    @property
+    def qubit_pairs(self) -> FrozenSet[Tuple['cirq.Qid', 'cirq.Qid']]:
+        """Returns the set of all couple-able qubits on the device."""
+        return self._qubit_pairs
+
+    @property
+    def gateset(self) -> 'cirq.Gateset':
+        """Returns the `cirq.Gateset` of supported gates on this device."""
+        return self._supported_gates
+
+    @property
+    def gate_durations(self) -> Optional[Dict['cirq.Gateset', 'cirq.Duration']]:
+        """Get a dictionary mapping from gateset to duration for gates."""
+        return self._gate_durations
+
+    def _value_equality_values_(self):
+        duration_equality = ''
+        if self._gate_durations is not None:
+            duration_equality = list(self._gate_durations.items())
+            duration_equality = sorted(duration_equality, key=lambda x: repr(x[0]))
+
+        return (
+            tuple(sorted(list(self._qubit_pairs))),
+            self._supported_gates,
+            tuple(duration_equality),
+        )
+
+    def _json_dict_(self):
+        duration_payload = ''
+        if self._gate_durations is not None:
+            duration_payload = list(self._gate_durations.items())
+            duration_payload = sorted(duration_payload, key=lambda x: repr(x[0]))
```

Comment:
`sorted` makes a list, so this can be done in one line. Example in `CircuitOperation`:
https://github.com/quantumlib/Cirq/blob/bfa2602876530840c9f95b2c3c405e332363ce37/cirq-core/cirq/circuits/circuit_operation.py#L349

Reply:
Done.